File: android/app/src/main/kotlin/com/wakeupsunshine/data/DebugLogger.kt
package com.wakeupsunshine.data

import android.content.Context
import android.util.Log
import com.wakeupsunshine.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URL
import java.net.HttpURLConnection
import java.text.SimpleDateFormat
import java.util.*

@Serializable
data class DebugEvent(
    val id: String = UUID.randomUUID().toString(),
    val timestamp: Long = System.currentTimeMillis(),
    val traceId: String? = null,
    val eventType: String,
    val message: String,
    val metadata: Map<String, String> = emptyMap(),
    val platform: String = "android"
)

object DebugLogger {
    private const val TAG = "WakeUpSunshine"
    private const val MAX_EVENTS = 100
    private const val PREFS_NAME = "debug_logger"
    private const val EVENTS_KEY = "debug_events"
    private const val LAST_WAKE_TRACE_KEY = "last_wake_trace_id"
    
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    
    @Volatile
    private var events: MutableList<DebugEvent> = mutableListOf()
    
    @Volatile
    var lastWakeTraceId: String? = null
        private set
    
    @Volatile
    var lastWakeResponse: Map<String, Any?>? = null
        private set
    
    @Volatile
    var lastNotificationPayload: Map<String, Any?>? = null
        private set
    
    @Volatile
    var lastAlarmTriggeredAt: Long? = null
        private set
    
    @Volatile
    var currentTraceId: String? = null
    
    private var context: Context? = null
    
    fun init(context: Context) {
        this.context = context.applicationContext
        loadEvents()
    }
    
    // MARK: - Core Logging
    
    fun log(
        eventType: String,
        message: String,
        traceId: String? = null,
        metadata: Map<String, String> = emptyMap()
    ) {
        val event = DebugEvent(
            traceId = traceId ?: currentTraceId,
            eventType = eventType,
            message = message,
            metadata = metadata
        )
        
        synchronized(events) {
            events.add(0, event)
            if (events.size > MAX_EVENTS) {
                events = events.take(MAX_EVENTS).toMutableList()
            }
        }
        
        saveEvents()
        
        // Log to Logcat
        val metadataStr = if (metadata.isEmpty()) "" else " | $metadata"
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "[$eventType] $message$metadataStr")
        }
    }

    fun debug(tag: String, message: String) {
        if (BuildConfig.DEBUG) {
            Log.d(tag, message)
        }
    }
    
    // MARK: - Convenience Methods
    
    fun logAppLaunch() {
        log("app_launch", "App launched")
    }
    
    fun logAuthRestored(userId: String, email: String?) {
        log("auth_restored", "Auth session restored", metadata = mapOf(
            "userId" to userId,
            "email" to (email ?: "unknown")
        ))
    }
    
    fun logContactsRefreshStart() {
        log("contacts_refresh_start", "Refreshing contacts")
    }
    
    fun logContactsRefreshComplete(
        peopleICanWake: Int,
        peopleWhoCanWakeMe: Int,
        pendingSent: Int,
        pendingReceived: Int
    ) {
        log("contacts_refresh_complete", "Contacts loaded", metadata = mapOf(
            "peopleICanWake" to peopleICanWake.toString(),
            "peopleWhoCanWakeMe" to peopleWhoCanWakeMe.toString(),
            "pendingSent" to pendingSent.toString(),
            "pendingReceived" to pendingReceived.toString()
        ))
    }
    
    fun logHomeDataLoaded(peopleICanWake: Int) {
        log("home_data_loaded", "Home data loaded", metadata = mapOf(
            "peopleICanWake" to peopleICanWake.toString()
        ))
    }
    
    fun logWakeButtonTapped(targetUserId: String, targetName: String) {
        currentTraceId = UUID.randomUUID().toString()
        log("wake_button_tapped", "Wake button tapped", traceId = currentTraceId, metadata = mapOf(
            "targetUserId" to targetUserId,
            "targetName" to targetName
        ))
    }
    
    fun logWakeRequestSent(traceId: String, targetUserId: String, alarmSoundId: String) {
        lastWakeTraceId = traceId
        saveLastWakeTraceId(traceId)
        
        log("wake_request_sent", "Wake request sent to backend", traceId = traceId, metadata = mapOf(
            "targetUserId" to targetUserId,
            "alarmSoundId" to alarmSoundId
        ))
    }
    
    fun logWakeResponse(response: Map<String, Any?>) {
        lastWakeResponse = response
        
        val traceId = response["traceId"] as? String
        val success = response["success"] as? Boolean ?: false
        val requestId = response["requestId"] as? String ?: "unknown"
        val devicesNotified = response["devicesNotified"] as? Int ?: 0
        
        log("wake_response", "Wake response received", traceId = traceId, metadata = mapOf(
            "success" to success.toString(),
            "requestId" to requestId,
            "devicesNotified" to devicesNotified.toString()
        ))
    }
    
    fun logFCMTokenGenerated(tokenPrefix: String, tokenSuffix: String) {
        log("fcm_token_generated", "FCM token generated", metadata = mapOf(
            "tokenPrefix" to tokenPrefix,
            "tokenSuffix" to tokenSuffix
        ))
    }
    
    fun logFCMTokenReceived(token: String) {
        val prefix = token.take(8)
        val suffix = token.takeLast(6)
        log("fcm_token_received", "FCM token received", metadata = mapOf(
            "tokenPrefix" to prefix,
            "tokenSuffix" to suffix
        ))
    }
    
    fun logDeviceTokenRegisterStart(token: String) {
        val prefix = token.take(8)
        val suffix = token.takeLast(6)
        log("device_token_register_start", "Registering device token with backend", metadata = mapOf(
            "tokenPrefix" to prefix,
            "tokenSuffix" to suffix
        ))
    }
    
    fun logDeviceTokenRegisterSuccess(deviceId: String) {
        log("device_token_register_success", "Device token registered successfully", metadata = mapOf(
            "deviceId" to deviceId
        ))
    }
    
    fun logDeviceTokenRegisterFailure(error: String) {
        log("device_token_register_failure", "Failed to register device token", metadata = mapOf(
            "error" to error
        ))
    }
    
    fun logFCMMessageReceived(payload: Map<String, Any?>, inForeground: Boolean) {
        lastNotificationPayload = payload
        
        val requestId = payload["request_id"] as? String ?: "unknown"
        val senderName = payload["sender_name"] as? String ?: "unknown"
        
        log("fcm_message_received", "FCM message received", metadata = mapOf(
            "requestId" to requestId,
            "senderName" to senderName,
            "inForeground" to inForeground.toString()
        ))
    }
    
    fun logAlarmActivityLaunched(requestId: String, senderName: String) {
        lastAlarmTriggeredAt = System.currentTimeMillis()
        
        log("alarm_activity_launched", "AlarmActivity launched", metadata = mapOf(
            "requestId" to requestId,
            "senderName" to senderName
        ))
    }
    
    fun logAlarmSoundPlaybackStarted(soundId: String) {
        log("alarm_sound_playback_started", "Playing alarm sound", metadata = mapOf(
            "soundId" to soundId
        ))
    }
    
    fun logAlarmSoundPlaybackFailed(error: String) {
        log("alarm_sound_playback_failed", "Failed to play alarm sound", metadata = mapOf(
            "error" to error
        ))
    }
    
    fun logError(error: Throwable, context: String) {
        log("error", "Error: $context", metadata = mapOf(
            "error" to (error.message ?: error.javaClass.simpleName)
        ))
    }
    
    // MARK: - Trace ID
    
    fun generateTraceId(): String {
        return UUID.randomUUID().toString()
    }
    
    // MARK: - Persistence
    
    private fun saveEvents() {
        context?.let { ctx ->
            try {
                val eventsJson = json.encodeToString(
                    kotlinx.serialization.builtins.ListSerializer(DebugEvent.serializer()),
                    events.toList()
                )
                ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    .edit()
                    .putString(EVENTS_KEY, eventsJson)
                    .apply()
            } catch (e: Exception) {
                Log.e(TAG, "Failed to save debug events", e)
            }
        }
    }
    
    private fun loadEvents() {
        context?.let { ctx ->
            try {
                val eventsJson = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    .getString(EVENTS_KEY, null)
                
                if (eventsJson != null) {
                    events = json.decodeFromString(
                        kotlinx.serialization.builtins.ListSerializer(DebugEvent.serializer()),
                        eventsJson
                    ).toMutableList()
                }
                
                lastWakeTraceId = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    .getString(LAST_WAKE_TRACE_KEY, null)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load debug events", e)
            }
        }
    }
    
    private fun saveLastWakeTraceId(traceId: String) {
        context?.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            ?.edit()
            ?.putString(LAST_WAKE_TRACE_KEY, traceId)
            ?.apply()
    }
    
    // MARK: - Clear
    
    fun clearEvents() {
        events.clear()
        lastWakeResponse = null
        lastNotificationPayload = null
        lastAlarmTriggeredAt = null
        
        context?.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            ?.edit()
            ?.clear()
            ?.apply()
    }
    
    // MARK: - Get Events
    
    fun getEvents(): List<DebugEvent> = synchronized(events) { events.toList() }
    
    // MARK: - Export
    
    fun exportDiagnostics(): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
        val sb = StringBuilder()
        
        sb.appendLine("=== Wake Up Sunshine Diagnostics ===")
        sb.appendLine("Generated: ${sdf.format(Date())}")
        sb.appendLine()
        
        sb.appendLine("--- Events (last ${events.size}) ---")
        events.forEach { event ->
            val metadataStr = if (event.metadata.isEmpty()) "" else " | ${event.metadata}"
            sb.appendLine("[${sdf.format(Date(event.timestamp))}] [${event.eventType}] ${event.message}$metadataStr")
        }
        
        lastWakeResponse?.let { response ->
            sb.appendLine()
            sb.appendLine("--- Last Wake Response ---")
            sb.appendLine(response.toString())
        }
        
        lastNotificationPayload?.let { payload ->
            sb.appendLine()
            sb.appendLine("--- Last FCM Payload ---")
            sb.appendLine(payload.toString())
        }
        
        lastAlarmTriggeredAt?.let { timestamp ->
            sb.appendLine()
            sb.appendLine("--- Last Alarm Triggered ---")
            sb.appendLine(sdf.format(Date(timestamp)))
        }
        
        return sb.toString()
    }
}

File: ios/WakeUpSunshine/Services/DebugLogger.swift
import Foundation
import SwiftUI

// MARK: - Debug Event
struct DebugEvent: Identifiable, Codable {
    let id: UUID
    let timestamp: Date
    let traceId: String?
    let eventType: String
    let message: String
    let metadata: [String: String]
    let platform: String
    
    init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        traceId: String? = nil,
        eventType: String,
        message: String,
        metadata: [String: String] = [:],
        platform: String = "ios"
    ) {
        self.id = id
        self.timestamp = timestamp
        self.traceId = traceId
        self.eventType = eventType
        self.message = message
        self.metadata = metadata
        self.platform = platform
    }
}

// MARK: - Debug Logger
class DebugLogger: ObservableObject {
    static let shared = DebugLogger()
    
    @Published var events: [DebugEvent] = []
    @Published var lastWakeTraceId: String?
    @Published var lastWakeResponse: [String: Any]?
    @Published var lastNotificationPayload: [String: Any]?
    @Published var lastAlarmTriggeredAt: Date?
    
    // Current trace ID for ongoing operations
    var currentTraceId: String?
    
    private let maxEvents = 100
    private let eventsKey = "debug_events"
    private let lastWakeTraceIdKey = "last_wake_trace_id"
    
    private init() {
        loadEvents()
    }
    
    // MARK: - Logging Methods
    
    func log(
        eventType: String,
        message: String,
        traceId: String? = nil,
        metadata: [String: String] = [:]
    ) {
        let event = DebugEvent(
            traceId: traceId ?? currentTraceId,
            eventType: eventType,
            message: message,
            metadata: metadata
        )
        
        DispatchQueue.mainAsyncSafe {
            self.events.insert(event, at: 0)
            
            // Keep only the last 100 events
            if self.events.count > self.maxEvents {
                self.events = Array(self.events.prefix(self.maxEvents))
            }
            
            self.saveEvents()
        }
        
        // Also print to console
        let metadataStr = metadata.isEmpty ? "" : " | \(metadata)"
        print("[DebugLogger] [\(eventType)] \(message)\(metadataStr)")
    }
    
    // MARK: - Specific Log Methods
    
    func logAppLaunch() {
        log(eventType: "app_launch", message: "App launched")
    }
    
    func logAuthRestored(userId: String, email: String?) {
        log(eventType: "auth_restored", message: "Auth session restored", metadata: [
            "userId": userId,
            "email": email ?? "unknown"
        ])
    }
    
    func logContactsRefreshStart() {
        log(eventType: "contacts_refresh_start", message: "Refreshing contacts")
    }
    
    func logContactsRefreshComplete(
        peopleICanWake: Int,
        peopleWhoCanWakeMe: Int,
        pendingSent: Int,
        pendingReceived: Int
    ) {
        log(eventType: "contacts_refresh_complete", message: "Contacts loaded", metadata: [
            "peopleICanWake": "\(peopleICanWake)",
            "peopleWhoCanWakeMe": "\(peopleWhoCanWakeMe)",
            "pendingSent": "\(pendingSent)",
            "pendingReceived": "\(pendingReceived)"
        ])
    }
    
    func logHomeDataLoaded(peopleICanWake: Int) {
        log(eventType: "home_data_loaded", message: "Home data loaded", metadata: [
            "peopleICanWake": "\(peopleICanWake)"
        ])
    }
    
    func logWakeButtonTapped(targetUserId: String, targetName: String) {
        currentTraceId = UUID().uuidString
        log(eventType: "wake_button_tapped", message: "Wake button tapped", traceId: currentTraceId, metadata: [
            "targetUserId": targetUserId,
            "targetName": targetName
        ])
    }
    
    func logWakeRequestSent(traceId: String, targetUserId: String, alarmSoundId: String) {
        lastWakeTraceId = traceId
        UserDefaults.standard.set(traceId, forKey: lastWakeTraceIdKey)
        
        log(eventType: "wake_request_sent", message: "Wake request sent to backend", traceId: traceId, metadata: [
            "targetUserId": targetUserId,
            "alarmSoundId": alarmSoundId
        ])
    }
    
    func logWakeResponse(_ response: [String: Any]) {
        lastWakeResponse = response
        
        let traceId = response["traceId"] as? String
        let success = response["success"] as? Bool ?? false
        let requestId = response["requestId"] as? String ?? "unknown"
        let devicesNotified = response["devicesNotified"] as? Int ?? 0
        
        log(eventType: "wake_response", message: "Wake response received", traceId: traceId, metadata: [
            "success": "\(success)",
            "requestId": requestId,
            "devicesNotified": "\(devicesNotified)"
        ])
    }
    
    func logAPNsRegistrationSuccess(tokenPrefix: String, tokenSuffix: String) {
        log(eventType: "apns_registration_success", message: "APNs token registered", metadata: [
            "tokenPrefix": tokenPrefix,
            "tokenSuffix": tokenSuffix
        ])
    }
    
    func logAPNsRegistrationFailure(error: String) {
        log(eventType: "apns_registration_failure", message: "APNs registration failed", metadata: [
            "error": error
        ])
    }
    
    func logDeviceTokenRegisterStart(token: String) {
        let prefix = String(token.prefix(8))
        let suffix = String(token.suffix(6))
        log(eventType: "device_token_register_start", message: "Registering device token with backend", metadata: [
            "tokenPrefix": prefix,
            "tokenSuffix": suffix
        ])
    }
    
    func logDeviceTokenRegisterSuccess(deviceId: String) {
        log(eventType: "device_token_register_success", message: "Device token registered successfully", metadata: [
            "deviceId": deviceId
        ])
    }
    
    func logDeviceTokenRegisterFailure(error: String) {
        log(eventType: "device_token_register_failure", message: "Failed to register device token", metadata: [
            "error": error
        ])
    }
    
    func logNotificationReceived(payload: [String: Any], inForeground: Bool) {
        lastNotificationPayload = payload
        
        let requestId = payload["requestId"] as? String ?? "unknown"
        let senderName = payload["senderName"] as? String ?? "unknown"
        
        log(eventType: "notification_received", message: "Push notification received", metadata: [
            "requestId": requestId,
            "senderName": senderName,
            "inForeground": "\(inForeground)"
        ])
    }
    
    func logWakeAlertPresented(requestId: String, senderName: String) {
        lastAlarmTriggeredAt = Date()
        
        log(eventType: "wake_alert_presented", message: "WakeAlert presented", metadata: [
            "requestId": requestId,
            "senderName": senderName
        ])
    }
    
    func logAlarmSoundPlaybackStarted(soundId: String) {
        log(eventType: "alarm_sound_playback_started", message: "Playing alarm sound", metadata: [
            "soundId": soundId
        ])
    }
    
    func logAlarmSoundPlaybackFailed(error: String) {
        log(eventType: "alarm_sound_playback_failed", message: "Failed to play alarm sound", metadata: [
            "error": error
        ])
    }
    
    func logError(_ error: Error, context: String) {
        log(eventType: "error", message: "Error: \(context)", metadata: [
            "error": error.localizedDescription
        ])
    }
    
    // MARK: - Generate Trace ID
    func generateTraceId() -> String {
        return UUID().uuidString
    }
    
    // MARK: - Persistence
    
    private func saveEvents() {
        guard let data = try? JSONEncoder().encode(events) else { return }
        UserDefaults.standard.set(data, forKey: eventsKey)
    }
    
    private func loadEvents() {
        guard let data = UserDefaults.standard.data(forKey: eventsKey),
              let loadedEvents = try? JSONDecoder().decode([DebugEvent].self, from: data) else {
            return
        }
        events = loadedEvents
        
        // Load last wake trace ID
        if let traceId = UserDefaults.standard.string(forKey: lastWakeTraceIdKey) {
            lastWakeTraceId = traceId
        }
    }
    
    // MARK: - Clear
    
    func clearEvents() {
        events = []
        lastWakeResponse = nil
        lastNotificationPayload = nil
        lastAlarmTriggeredAt = nil
        UserDefaults.standard.removeObject(forKey: eventsKey)
        UserDefaults.standard.removeObject(forKey: lastWakeTraceIdKey)
    }
    
    // MARK: - Export
    
    func exportDiagnostics() -> String {
        var export = "=== Wake Up Sunshine Diagnostics ===\n"
        export += "Generated: \(ISO8601DateFormatter().string(from: Date()))\n\n"
        
        export += "--- Events (last \(events.count)) ---\n"
        for event in events {
            let metadataStr = event.metadata.isEmpty ? "" : " | \(event.metadata)"
            export += "[\(event.timestamp)] [\(event.eventType)] \(event.message)\(metadataStr)\n"
        }
        
        if let lastWake = lastWakeResponse {
            export += "\n--- Last Wake Response ---\n"
            export += "\(lastWake)\n"
        }
        
        if let lastNotif = lastNotificationPayload {
            export += "\n--- Last Notification Payload ---\n"
            export += "\(lastNotif)\n"
        }
        
        if let lastAlarm = lastAlarmTriggeredAt {
            export += "\n--- Last Alarm Triggered ---\n"
            export += "\(lastAlarm)\n"
        }
        
        return export
    }
}

// MARK: - Helper Extension
extension DispatchQueue {
    static func mainAsyncSafe(execute work: @escaping () -> Void) {
        if Thread.isMainThread {
            work()
        } else {
            DispatchQueue.main.async(execute: work)
        }
    }
}